pyfr/plugins/acoustics.py

```diff
@@ -0,0 +1,180 @@
+
+from pyfr.mpiutil import get_comm_rank_root
+from pyfr.nputil import npeval
+from pyfr.plugins.base import BaseSolnPlugin, init_csv, RegionMixin
+
+
+class FWHPlugin(RegionMixin, BaseSolnPlugin):
+    name = 'fwhsolver'
+    systems = ['ac-euler', 'ac-navier-stokes', 'euler', 'navier-stokes']
+    formulations = ['dual', 'std']
+
+    def __init__(self, intg, cfgsect, suffix=None, *args, **kwargs):
+        super().__init__(intg, cfgsect, suffix)
+        comm, rank, root = get_comm_rank_root()
+
+        self.elementscls = intg.system.elementscls
+
+        self.tstart = self.cfg.getfloat(cfgsect, 'tstart', 0.0)
+        self.dt = self.cfg.getfloat(cfgsect, 'dt')
+        self.t_window = self.cfg.getfloat(cfgsect, 'tend', intg.tend)
+        self.obsv_pts = np.array(self.cfg.getliteral(self.cfgsect,
+                                                     'observer-pts'))
+        self.nobvs = self.obsv_pts.shape[0]
+        ndims = self.ndims
+        self.t_last = -np.inf
+
+        # Initialise data file
+        if rank == root:
+            header = ','.join(['t', 'x', 'y', 'z'][:self.ndims + 1] + ['mag'])
+            self.outf = init_csv(self.cfg, cfgsect, header)
+
+        # Far field condition
+        self.incomp = intg.system.name in {'ac-euler', 'ac-navier-stokes'}
+        privarmap = intg.system.elementscls.privarmap[ndims]
+        self._vidx = [x in 'uvw' for x in privarmap]
+        self._pidx = privarmap.index('p')
+        self.consts = self.cfg.items_as('constants', float)
+
+        self.qinf = {k: npeval(self.cfg.getexpr(cfgsect, k), self.consts)
+                     for k in privarmap}
+        self.uinf = np.array([self.qinf[k] for k in 'uvw'[:ndims]])
+
+        if self.incomp:
+            self.qinf['rho'] = self.uinf['p']/self.consts['ac-zeta']
+            self.qinf['c'] = np.sqrt(self.consts['ac-zeta'])
+        else:
+            gamma = self.consts['gamma']
+            self.qinf['c'] = np.sqrt(gamma * self.qinf['p'] / self.qinf['rho'])
+        self.qinf['M'] = [self.qinf[k]/self.qinf['c'] for k in 'uvw'[:ndims]]
+        self.Minf = np.linalg.norm(self.qinf['M'], ord=2)
+
+        ele_map = intg.system.ele_map
+        self.emap = {k: i for i, k in enumerate(ele_map.keys())}
+        self._init_surf(ele_map)
+
+    def _init_surf(self, ele_map):
+        ndims = self.ndims
+
+        self.proj = {}
+
+        for etype, v in self._surf_region_data.items():
+            eles = ele_map[etype]
+
+            pnorm = eles._pnorm_fpts
+            rcpdjacs = eles.rcpdjac_at_np('fpts')
+
+            for fidx in set(x for i, x in v):
+                eidx = [i for i, x in v if x == fidx]
+                facefpts = eles.basis.facefpts[fidx]
+                fslice = np.ix_(facefpts, eidx)
+                npts = len(eidx)*len(facefpts)
+
+                m0 = eles.basis.m0[facefpts]
+                a = eles.basis.fpts_wts[facefpts][:, None] / rcpdjacs[fslice]
+                a = np.reshape(a, (-1,))
```

Comment:
Prefer `a.reshape` (the module level functions are used when we are not sure that `a` is a NumPy array).